lmdeploy/pytorch/paging/block_trie.py

```diff
@@ -0,0 +1,215 @@
+        self.hash_key = hash_key
+        self.block = block
+        self.tokens = tokens
+        self.num_matched = num_matched
+        self.children: Dict[int, 'Node'] = dict()
+        self._parent: 'Node' = None
+
+    @property
+    def parent(self):
+        return self._parent
+
+    @parent.setter
+    def parent(self, val: 'Node'):
+        old_parent = self._parent
+        if old_parent is not None:
+            old_parent.children.pop(self.hash_key)
+        if val is not None:
+            val.children[self.hash_key] = self
+        self._parent = val
+
+    def __lt__(self, other):
+        return True
+
+    def __le__(self, other):
+        return True
+
+
+class BlockTrie:
+
+    def __init__(self, cache_config: CacheConfig,
+                 block_manager: BaseBlockManager):
+        self.block_manager = block_manager
+        self.cache_config = cache_config
+        self.allocator = self.block_manager.allocator
+        self.block_size = cache_config.block_size
+        self.enable = self.cache_config.prefix_caching
+
+        # caches with different adapter should not be shared.
+        self._roots: Dict[str, Node] = dict()
+        self.leaves: Set[Node] = set()
+
+    def get_root(self, adapter_name: str):
+        if adapter_name not in self._roots:
+            self._roots[adapter_name] = Node(-1, -1, None)
+        return self._roots[adapter_name]
+
+    def match(self, seq: SchedulerSequence):
+        if not self.enable:
+            return
+
+        block_size = self.block_size
+        matched_blocks = []
+
+        logical_blocks = seq.logical_blocks
+        curr: Node = getattr(logical_blocks, 'last_shared_node', None)
+        if curr is None:
+            curr = self.get_root(seq.adapter_name)
+        num_matched = curr.num_matched
+
+        def __match_success(node: Node):
+            nonlocal curr, num_matched
+            matched_blocks.append(node.block)
+            curr = node
+            num_matched += block_size
+
+        while num_matched + block_size < seq.num_all_ids:
+            curr_tokens = seq.history_cache[num_matched:num_matched +
+                                            block_size]
+
+            key = hash(tuple(curr_tokens))
+            if key not in curr.children:
+                break
+
+            child = curr.children[key]
+            if not np.array_equal(curr_tokens, child.tokens):
+                break
+
+            __match_success(child)
+
+        if len(matched_blocks) > 0:
+            matched_blocks = np.array(matched_blocks)
+            self.allocator.update_access_time(matched_blocks)
+            self.allocator.add_ref_count(matched_blocks, 1)
+            seq.logical_blocks.append(matched_blocks)
+            seq.set_step(num_matched)
+
+        seq.logical_blocks.last_shared_node = curr
+
+    def allocate(self, seq: SchedulerSequence, prealloc_size: int = 0):
+        self.block_manager.allocate(seq, prealloc_size)
```

Comment:
The `block_manager.allocate` can be moved out of this function to decouple the `caching` and `block allocation`.
This function is mainly responsible for caching/inserting the computed sequence blocks. In `_schedule_prefill`, we only need to allocate and don't need to cache.

Reply:
Sure, sounds reasonable.

Honestly, I am confused about the design of paging modules. block_trie requires `free` to evict blocks, and I think `allocate` and `free` should be placed in the very same module, that's why I wrap block_manager in block_trie.

Reply:
I thought the `block_trie` is responsible for adding/removing node and increasing/decreasing the block ref count. The `block_manager` is responsible for allocating/freeing blocks.